### ISSUE TITLE
A given test suite should only compile components once?

### ISSUE THREAD

**Issue by brian428** [angular/angular#13963]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Looking at the code for TestBed, it appears that `compileComponents()` will recompile all classes specified by `configureTestingModule()` before _every_ test spec. `compileComponents()` checks `this._instantiated`, and if true it immediately resolves the Promise. But `configureTestingModule()` uses `this._assertNotInstantiated()`, which ensures that `this._instantiated` is false. So the compile run seems to happen before every spec.

**Expected behavior**
I would think that for a given suite and it's TestBed configuration, the compilation should only happen once? Obviously all objects should be _recreated_ between specs, but I don't understand the logic of _recompiling_ everything more than once.

**What is the motivation / use case for changing the behavior?**
The compilation can take significant time to perform. Doing it once per suite would seem to provide a big speed increase, especially for large numbers of suites or suites with lots of specs.

* **Angular version:** 2.4.3

* **Browser:** all 

* **Language:** TypeScript 2.1.5

**Comment by DzmitryShylovich** [comment]
Workaround is to use webpack. webpackdevelopers need not call compileComponents because it inlines templates and css as part of the automated build process that precedes running the test. 😄

**Comment by brian428** [comment]
Well, thanks for the comment, but "use a different module loader" doesn't help with the underlying problem at all. 

I assume TestBed could just be changed to only compile once and cache the results for that suite, which would mean that subsequent calls should resolve immediately. The only thing I'm not sure of is whether there's some _reason_ why the compilation results aren't reused.

**Comment by brian428** [comment]
Any update on this? It just seems bizarre (not to mention terribly inefficient) to _recompile_ all the components over and over for every spec method.

Obviously, components need to be _recreated_ between spec methods. But I see no possible reason why the components would need to be _recompiled_ more than once for a given spec.

**Comment by tbosch** [comment]
Agree. However, this is not trivial if a component / directive / pipe was modified via `TestBed.overrideXXX`. Then we have to recompile components in the same `NgModule`, and all components of `NgModule`s that import it. This is needed as e.g. the selector of a component could have changed.

A simple way of speeding things up is to have a fast path for when no component has been overwritten.

**Comment by brian428** [comment]
Thanks, @tbosch, though I admit I'm not totally clear on what you mean about using `TestBed.overrideXXX`. Typically, even when I use `overrideXXX`, it's done at the same point everything else about the testing module is being configured. E.g.:

```
TestBed.configureTestingModule(...)
  .overrideModule(...)
  .overrideComponent(...)
  .compileComponents()
  .then(...)
```

The above seems like it is just as much a candidate for caching/reusing the compile result. 

So maybe you're talking about some less common way of using the `overrideXXX` methods? If so, maybe provide a way for us to indicate that we want to cache/reuse the compile result even if `overrideXXX` is used?

Either something like:
```
TestBed.configureTestingModule(...)
  .overrideModule(...)
  .overrideComponent(...)
  .compileComponents( {cache:true} )
  .then(...)
```

or possibly:

```
TestBed.configureTestingModule(...)
  .overrideModule(...)
  .overrideComponent(...)
  .compileComponents()
  .cache()
  .then(...)
```

Basically, I think the cached compile results should still be an option even if `overrideXXX` is used, as long as the developer indicates that the test module configuration isn't changing between spec runs.

Let me know if that makes sense. Thanks!

**Comment by ronnyek** [comment]
I utilized this little helper: https://github.com/angular/angular/issues/12409#issuecomment-322832870
Took my tests running in ~2min48sec all the way down to 6.2sec.

Not to say it will work in every situation, but MOST of those 100 unit tests required very little to be converted over.

**Comment by brian428** [comment]
The key issue with that approach is that it doesn't recreate the dependencies between tests, so any state held by the dependencies can cause problems. The ideal option is to just compile to test module once and reuse that, since the vast majority of the time is spent compiling the test module.

[This project](https://github.com/Quramy/ngx-zombie-compiler) does something close. But I admit I'm still not sure what the delay is in getting Angular itself to support this. I get that using the `overrideX()` methods can interfere, but as long as the test module is defined once and no overrides are used after it is created, this is really not all that complicated.

**Comment by brian428** [comment]
Any update on this from the Angular team? I've hacked together a compiler implementation that only compiles templates once, and it reduces the execution time for my tests by 50-75%. It shaves 5+ minutes off _every_ test run. That's a huge difference when I'm running 700+ tests dozens of times a day. 

However, the compiler API changed with v5, so now I have to try and re-hack this all over again. Can this issue be given some priority please?

**Comment by Koslun** [comment]
@brian428 A question, that was basically this issue, was raised during the QA session of the last Angular Connect (november 2017). It's at 2:34:56in this [stream](https://youtu.be/exENideeTDg?t=2h34m56s). Basically Igor Minar answered that incremental compilation was introduced in version 5 and that they know there are some more things they can do with the test bed and that this will be worked on in version 6 and 7.

So interpret that as we can expect it some time next year and this issue is in fact on their radar and current plans.

**Comment by brian428** [comment]
@eric-simonton-sama after spending some time on it, I wasn't able to figure it out. I may take another stab at it later.

While working on this, I also discovered the `aotSummaries` arg for `initTestEnvironment()` and `configureTestingModule()`. So it seems like all of the pieces are there internally to allow reusing the compiled templates in subsequent spec methods in the same spec. TestBed just needs to provide some way to tell it to reuse the summaries instead of clearing everything out between spec methods. 

Or, a less ideal option would be to provide some way to get at the current set of aotSummaries after `compileComponents()` is called, so we can pass them back in via the `aotSummaries` arg for subsequent spec methods. Individual summaries can be pulled from the `CompileMetadataResolver`, but manually pulling out every summary individually (by type) isn't a viable option.

**Comment by ersimont** [comment]
@brian428 I finally spent a day focused on this issue. I published my results in the [`precompileForTests()`](https://simontonsoftware.github.io/s-ng-dev-utils/typedoc/index.html#precompilefortests) function in [`s-ng-dev-utils`](https://github.com/simontonsoftware/s-ng-dev-utils). So far it seems to be working well.

It works exactly as you suggest, calling `compileComponents()` extracting their `aotSummaries`, then passing them in to a refreshed `initTestEnvironment()`. The test environment is reused between tests (unlike `configureTestingModule()`).

**Comment by FrancescoBorzi** [comment]
are there any news about this?

**Comment by moscoso** [comment]
I don't understand how Angular can pride itself on its test-driven development angle if this is still an issue years later... It took me hours to realize why I couldn't put test-suite setup code in the beforeAll() method. 

How can I have a sane developer TDD experience when it takes several minutes to run specs for my angular method. 

At the very least there needs to be a mention of this horrible Testbed compiler side effect in the documentation because on the surface, there is no distinct error message or explanation for why tests can't be set up in the beforeAll() method.

**Comment by JoostK** [comment]
Ivy's `TestBed` only recompiles components for which an override had been registered, otherwise it's able to reuse prior compilations (even AOT compiled components for libs).
